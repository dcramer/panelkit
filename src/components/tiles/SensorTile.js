import PropTypes from "prop-types";

import Tile, { TileProps } from "../Tile";

export default class SensorTile extends Tile {
  static propTypes = {
    ...TileProps,
    entityId: PropTypes.string.isRequired,
    format: PropTypes.function,
  };

  static defaultProps = {
    format: (state, attributes) => `${state} ${attributes.unit_of_measurement}`,
  };

  getWatchedEntityIds() {
    return [this.props.entityId];
  }

  renderBody() {
    const { state, attributes } = this.getEntity(this.props.entityId);
    return this.props.format(state, attributes);
  }

  renderStatus() {}
}
