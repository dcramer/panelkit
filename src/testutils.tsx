import React from "react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { Entity } from "./types";
import HomeAssistant, { SuggestedChanges } from "./hass";

export const URL = "http://localhost:8123";

export const ACCESS_TOKEN = "example.access.token";

export const DEFAULT_CONFIG = {
  url: URL,
  accessToken: ACCESS_TOKEN,
  tiles: [],
};

export const simulateTouch = (container: any) => {
  userEvent.click(container);
};

export const getTileProps = (entity: Entity) => {
  const hass = new HomeAssistant(DEFAULT_CONFIG);

  hass.callService = jest
    .fn()
    .mockImplementation(
      (
        domain: string,
        service: string,
        serviceData: any,
        suggestedChanges: SuggestedChanges | null = null
      ) => {
        return Promise.resolve();
      }
    );
  hass.sendMessage = jest.fn().mockImplementation(() => {
    return Promise.resolve();
  });

  hass.populateEntityCache([entity]);

  return {
    entityId: entity.entity_id,
    hass,
    cameraList: [],
  };
};

export const renderTile = (
  Component: any,
  entity: Entity,
  props: { [key: string]: any } = {}
) => {
  const tileProps = {
    ...getTileProps(entity),
    ...props,
  };

  const result = render(<Component {...tileProps} />);

  return {
    ...result,
    hass: tileProps.hass,
    container: result.container.firstChild,
  };
};
