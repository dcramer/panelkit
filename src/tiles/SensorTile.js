import React from "react";
import PropTypes from "prop-types";

import Tile, { TileProps } from "./Tile";

export default class SensorTile extends Tile {
  static propTypes = {
    ...TileProps,
    entityId: PropTypes.string.isRequired,
    format: PropTypes.function,
    unit: PropTypes.string,
  };

  static defaultProps = {
    format: (state, _attributes, unitOfMeasurement = null) => (
      <React.Fragment>
        {state} <span className="unit">{unitOfMeasurement}</span>
      </React.Fragment>
    ),
  };

  renderBody() {
    const { state, attributes } = this.getEntity(this.props.entityId);
    return this.props.format(
      state,
      attributes,
      this.props.unit || attributes.unit_of_measurement
    );
  }

  renderStatus() {}
}
