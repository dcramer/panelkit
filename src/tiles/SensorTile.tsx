import React from "react";

import Tile, { TileProps } from "./Tile";
import { EntityAttributes } from "../types";

type SensorTileProps = TileProps & {
  entityId: string;
  format: Function;
  unit?: string;
};

type SensorAttributes = EntityAttributes & {
  unit_of_measurement?: string;
};

export default class SensorTile extends Tile<SensorTileProps> {
  static defaultProps = {
    format: (
      state: string,
      attributes: EntityAttributes,
      unitOfMeasurement: string | null = null
    ) => (
      <React.Fragment>
        {state}
        <span className="unit">{unitOfMeasurement}</span>
      </React.Fragment>
    ),
  };

  getUnit(entityId: string) {
    if (this.props.unit) return this.props.unit;
    const {
      attributes: { unit_of_measurement },
    } = this.getEntity(entityId);
    return unit_of_measurement || "";
  }

  renderBody() {
    const { state, attributes } = this.getEntity(this.props.entityId);
    return (
      <div className="tile-sensor">
        {this.props.format(
          state,
          attributes,
          this.getUnit(this.props.entityId)
        )}
      </div>
    );
  }

  renderStatus() {}
}
