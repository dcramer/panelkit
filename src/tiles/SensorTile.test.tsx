import { renderTile } from "../testutils";
import SensorTile from "./SensorTile";

test("renders", async () => {
  const { container, getByText } = renderTile(SensorTile, {
    entity_id: "sensor.test",
    state: "73",
    attributes: {
      unit_of_measurement: "F",
      friendly_name: "Test Sensor",
    },
  });

  expect(getByText(/Test Sensor/i)).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});
