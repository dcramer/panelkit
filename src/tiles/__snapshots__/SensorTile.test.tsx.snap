// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  class="tile state-73 false"
  style="cursor: normal;"
>
  <div
    class="tile-container"
  >
    <div
      class="tile-body"
    >
      <div
        class="tile-sensor"
      >
        73
        <span
          class="unit"
        >
          F
        </span>
      </div>
    </div>
    <div
      class="tile-title"
    >
      Test Sensor
    </div>
  </div>
</div>
`;
